File: macos-app/TodoSync/Sources/TodoSync/SyncService.swift
import Foundation
import EventKit

@MainActor
class SyncService {
    let api: ApiClient
    let db: Database
    let eventKit: EventKitService
    var ws: WebSocketClient?
    var deviceInfo: DeviceInfo?
    let deviceId: String
    var pollingTimer: Timer?

    let defaults = UserDefaults.standard
    let deviceIdKey = "todo_sync_device_id"
    let tokenKey = "todo_sync_token"

    init(baseURL: String) {
        self.api = ApiClient(baseURL: baseURL)
        self.db = Database()
        self.eventKit = EventKitService()
        self.deviceId = UUID().uuidString
    }

    func setup() async {
        if let savedId = defaults.string(forKey: deviceIdKey),
           let savedToken = defaults.string(forKey: tokenKey) {
            deviceInfo = DeviceInfo(id: savedId, token: savedToken, expiresAt: "")
            api.token = savedToken
        } else {
            guard let info = try? await api.register() else { return }
            deviceInfo = info
            api.token = info.token
            defaults.set(info.id, forKey: deviceIdKey)
            defaults.set(info.token, forKey: tokenKey)
        }

        if let url = URL(string: api.baseURL
            .replacingOccurrences(of: "http", with: "ws") + "/ws") {
            ws = WebSocketClient(url: url, token: api.token ?? "")
            ws?.onMessage = { [weak self] msg in
                Task { @MainActor in self?.handlePush(msg) }
            }
            ws?.connect()
        }

        startPolling()
    }

    private func startPolling() {
        pollingTimer = Timer.scheduledTimer(withTimeInterval: 30, repeats: true) { [weak self] _ in
            Task { @MainActor in await self?.syncFromEventKit() }
        }
        Task { @MainActor in await syncFromEventKit() }
    }

    func syncFromEventKit() async {
        guard await eventKit.requestPermission() else { return }

        let reminders = eventKit.fetchAllReminders()
        let lastSync = db.getLastSyncTime()

        for reminder in reminders {
            let calId = reminder.calendarItemIdentifier
            let modDate = reminder.lastModifiedDate ?? Date.distantPast
            let modDateStr = ISO8601DateFormatter().string(from: modDate)

            if modDateStr <= lastSync { continue }

            if let localId = db.getLocalId(calendarItemId: calId) {
                let existing = db.getTodo(localId)
                if existing?.updatedAt ?? "" < modDateStr {
                    let todo = makeTodo(id: localId, reminder: reminder, modDateStr: modDateStr)
                    db.insertTodo(todo)
                    try? await api.pushLocalChanges(todos: [todo], deviceId: deviceId)
                }
            } else {
                let localId = "mac-\(UUID().uuidString)"
                db.saveMapping(localId: localId, calendarItemId: calId)
                let todo = makeTodo(id: localId, reminder: reminder, modDateStr: modDateStr)
                db.insertTodo(todo)
                try? await api.createTodo(todo)
            }
        }
    }

    private func makeTodo(id: String, reminder: EKReminder, modDateStr: String) -> Todo {
        Todo(
            id: id,
            title: reminder.title ?? "",
            notes: reminder.notes ?? "",
            dueDate: nil,
            isCompleted: reminder.isCompleted,
            priority: reminder.priority,
            updatedAt: modDateStr,
            deletedAt: nil,
            source: "macos"
        )
    }

    func syncFromServer() async {
        let since = db.getLastSyncTime()
        guard let todos = try? await api.fetchTodos(since: since) else { return }

        for todo in todos {
            if todo.source == "macos" { continue }
            db.insertTodo(todo)

            if todo.deletedAt != nil {
                if let calId = db.getCalendarItemId(localId: todo.id),
                   let reminder = eventKit.getReminder(byIdentifier: calId) {
                    eventKit.deleteReminder(reminder)
                }
                continue
            }

            if let calId = db.getCalendarItemId(localId: todo.id),
               let reminder = eventKit.getReminder(byIdentifier: calId) {
                eventKit.updateReminder(reminder, from: todo)
            } else {
                let reminder = eventKit.createReminder(from: todo)
                db.saveMapping(localId: todo.id, calendarItemId: reminder.calendarItemIdentifier)
            }
        }
    }

    func handlePush(_ msg: PushMessage) {
        if msg.data.source == "macos" { return }
        db.insertTodo(msg.data)

        switch msg.type {
        case "todo.deleted":
            if let calId = db.getCalendarItemId(localId: msg.data.id),
               let reminder = eventKit.getReminder(byIdentifier: calId) {
                eventKit.deleteReminder(reminder)
            }
        case "todo.created":
            let reminder = eventKit.createReminder(from: msg.data)
            db.saveMapping(localId: msg.data.id, calendarItemId: reminder.calendarItemIdentifier)
        case "todo.updated":
            if let calId = db.getCalendarItemId(localId: msg.data.id),
               let reminder = eventKit.getReminder(byIdentifier: calId) {
                eventKit.updateReminder(reminder, from: msg.data)
            } else {
                let reminder = eventKit.createReminder(from: msg.data)
                db.saveMapping(localId: msg.data.id, calendarItemId: reminder.calendarItemIdentifier)
            }
        default:
            break
        }
    }
}

File: harmony-app/entry/src/main/ets/service/SyncService.ts
import { Todo, PushMessage } from '../model/Todo'
import { ApiClient } from '../model/ApiClient'
import { WebSocketClient } from '../model/WebSocketClient'
import { Database } from '../database/Database'

export class SyncService {
  private api: ApiClient
  private db: Database
  private ws: WebSocketClient
  private token: string = ''
  private deviceId: string

  constructor(baseURL: string, context: Context) {
    this.api = new ApiClient(baseURL)
    this.db = new Database()
    this.deviceId = 'hm-' + this.generateId()
    this.ws = new WebSocketClient(baseURL, '')
  }

  private generateId(): string {
    return Date.now().toString(36) + Math.random().toString(36).slice(2, 8)
  }

  async setup(context: Context): Promise<void> {
    await this.db.init(context)
    const prefs = context.getApplicationContext().getPreferences('todo_sync_prefs')
    const savedToken = prefs.get('token', '')
    const savedId = prefs.get('device_id', '')

    if (savedToken && savedId) {
      this.token = savedToken as string
      this.api.setToken(this.token)
    } else {
      const info = await this.api.register()
      this.token = info.token
      this.api.setToken(this.token)
      prefs.put('token', this.token)
      prefs.put('device_id', info.id)
      prefs.flush()
    }

    this.setupWebSocket()

    this.syncFromServer()
  }

  private setupWebSocket() {
    this.ws.onMessage = (msg: PushMessage) => {
      this.handlePush(msg)
    }
    this.ws.connect()
  }

  async syncFromServer(): Promise<void> {
    const since = await this.db.getLastSyncTime()
    try {
      const todos = await this.api.fetchTodos(since)
      for (const todo of todos) {
        if (todo.source === 'harmony') continue
        await this.db.insertTodo(todo)
      }
    } catch (e) {
      console.error('syncFromServer error:', e)
    }
  }

  async createTodo(title: string, notes: string): Promise<Todo> {
    const id = 'hm-' + this.generateId()
    const now = new Date().toISOString()
    const todo = new Todo(id, title, notes, '', false, 0, now, '', 'harmony')

    await this.db.insertTodo(todo)
    try {
      await this.api.createTodo(todo)
    } catch (e) {
      console.error('createTodo push error:', e)
    }
    return todo
  }

  async updateTodo(todo: Todo): Promise<void> {
    const now = new Date().toISOString()
    const updated = new Todo(todo.id, todo.title, todo.notes, todo.dueDate,
      todo.isCompleted, todo.priority, now, todo.deletedAt, 'harmony')

    await this.db.insertTodo(updated)
    try {
      const fields: Record<string, Object> = {
        title: { value: todo.title, updatedAt: now, deviceId: this.deviceId },
        notes: { value: todo.notes, updatedAt: now, deviceId: this.deviceId },
        isCompleted: { value: todo.isCompleted, updatedAt: now, deviceId: this.deviceId }
      }
      await this.api.updateTodo(todo.id, fields)
    } catch (e) {
      console.error('updateTodo push error:', e)
    }
  }

  async deleteTodo(id: string): Promise<void> {
    await this.db.insertTodo(
      new Todo(id, '', '', '', false, 0, new Date().toISOString(),
        new Date().toISOString(), 'harmony')
    )
    try {
      await this.api.deleteTodo(id)
    } catch (e) {
      console.error('deleteTodo push error:', e)
    }
  }

  async getTodos(): Promise<Todo[]> {
    return await this.db.getAllTodos()
  }

  private handlePush(msg: PushMessage) {
    if (msg.data.source === 'harmony') return
    this.db.insertTodo(msg.data)
    if (msg.type === 'todo.deleted') {
      this.notifyDataChange()
    }
  }

  // Used to trigger UI refresh from push
  private onChangeCallback: (() => void) | null = null
  set onDataChange(callback: () => void) {
    this.onChangeCallback = callback
  }
  private notifyDataChange() {
    this.onChangeCallback?.()
  }
}
